include/fiction/algorithms/simulation/sidb/clustercomplete.hpp

```diff
@@ -0,0 +1,476 @@
+
+        // swap back
+        std::swap(clustering_state.back(), clustering_state[max_cst_ix]);
+    }
+
+    void collect_physically_valid_charge_distributions(const sidb_cluster_ptr& top_cluster) noexcept
+    {
+        const uint64_t top_level_multisets = top_cluster->charge_space.size();
+
+        const uint64_t num_threads_to_use = std::min(std::max(num_threads, uint64_t{1}), top_level_multisets);
+
+        // define the bit string ranges per thread
+        std::vector<std::pair<uint64_t, uint64_t>> ranges;
+        const uint64_t chunk_size = std::max(top_level_multisets / num_threads_to_use, uint64_t{1});
+        uint64_t       start      = 0;
+        uint64_t       end        = chunk_size - 1;
+
+        for (uint64_t i = 0; i < num_threads_to_use; ++i)
+        {
+            ranges.emplace_back(start, end);
+            start = end + 1;
+            end   = i == num_threads_to_use - 2 ? top_level_multisets - 1 : start + chunk_size - 1;
+        }
+
+        std::vector<std::thread> threads_vec{};
+        threads_vec.reserve(num_threads_to_use);
+
+        for (const std::pair<uint64_t, uint64_t>& range : ranges)
+        {
+            threads_vec.emplace_back(
+                [&]
+                {
+                    // iterate over all multiset charge configurations in the assigned range
+                    for (uint64_t ix = range.first; ix <= range.second; ++ix)
+                    {
+                        // iterate over all cluster charge assignments in the multiset charge configuration
+                        for (sidb_cluster_state_composition& composition :
+                             std::next(top_cluster->charge_space.cbegin(), static_cast<int64_t>(ix))->compositions)
+                        {
+                            sidb_clustering_state clustering_state{};
+                            clustering_state.reserve(charge_layout.num_cells());
+
+                            for (sidb_cluster_state& cst : composition)
+                            {
+                                clustering_state.emplace_back(std::make_unique<sidb_cluster_state>(std::move(cst)));
+                            }
+
+                            add_physically_valid_charge_confs(clustering_state);
+                        }
+                    }
+                });
+        }
+
+        for (std::thread& t : threads_vec)
+        {
+            t.join();
+        }
+    }
+
+    sidb_simulation_result<Lyt> res{};
+    std::mutex                  res_mutex{};
+
+    const charge_distribution_surface<Lyt> charge_layout;
+    const std::array<double, 4>            mu_bounds_with_error;
+    const uint64_t                         num_threads;
+};
+
+}  // namespace detail
+
+/**
+ * *ClusterComplete* is an *exact* simulator that radicalizes the SiDB simulation domain with its potential. SiDB logic
+ * simulation in *base 3* with a 100% certainty of accuracy is made available to the domain for well over 50 SiDBs. It
```

Comment:
```suggestion
 * simulation in *base 3* with a 100% certainty of accuracy is made available to the domain for over 50 SiDBs. It
```

Reply:
"well over" is perhaps a bit informal, but 50 is also a timid statement, seeing that I could simulate a CX with PIs and POs at 79 DBs.

Reply:
In order to create a better statement I'm running benchmarks for a little while now from which I can extract a performance scaling over layout _density_. Then, computing an average density for (big) Bestagon conglomeration as SiDB logic representative I can extract a scaling for SiDB logic simulation. I'm expecting the density for Bestagon gate chains to be really small compared to the random layouts I'm benchmarking in enclosed spaces.
The little bit tricky part here is that Bestagon gates have only a few top level multiset configurations that are valid (because of BDL principles) but there are many combination paths to look in for these few (all kinds of kinks etc). With how multithreading is right now, that means that one core will be going for a considerable time on its own, compared to random layouts that give a more even thread load due to a good variation in top level multiset configurations. In the end however, such MT optimisations are (roughly) linear, so it should not matter for a statement on scaling.